raw_bsd.go

```diff
@@ -192,6 +200,73 @@ func (p *packetConn) SetPromiscuous(b bool) error {
 	return syscall.SetBpfPromisc(p.fd, m)
 }
 
+// ifReq is the BSD struct passed to ioctl for multicast add/drop operations.
+type ifReq struct {
+	ifrName [unix.IFNAMSIZ]uint8
+	ifrAddr unix.SockaddrDatalink
+}
+
+const etherAdderAlen = 6
+
+// JoinMulticast joins or drops membership of data link layer multicast address on interface,
+// allowing it to recieve multicast traffic.
+func (p *packetConn) SetMulticast(b bool, addr net.HardwareAddr) error {
+	// Convert hardware address back to byte array form
+	var baddr [8]byte
+	copy(baddr[:], addr)
+
+	s, err := syscall.Socket(unix.AF_INET, unix.SOCK_DGRAM, unix.IPPROTO_UDP)
```

Comment:
My limited understanding of the BSD API; in this case the socket is really just for dummy purposes for the future ioctl call. So long as it is in the correct family I believe it should be fine. I'm going to take some notes on the implementations I can find from BSD family OS's and either include it in comments or document it elsewhere. The main issue has been finding documentation, not many examples in the wild of adding L2 addresses to interface mcast groups. Most examples revolve around L3 addresses and generation of IGMP. For all this testing I am using port auth (802.1x) class addresses (01:80:C2:00:00:03) so they are routed over L2 domain without any other action required.

Example of how this is accomplished on FreeBSD in mtest source: https://github.com/freebsd/freebsd/blob/master/usr.sbin/mtest/mtest.c#L759
Example of how the above likely needs modification for OpenBSD/NetBSD/Solaris: https://github.com/vincentbernat/lldpd/blob/08203be9bfd915ebc12f266bcd6adf8391cc5ef5/src/daemon/priv.c#L259

One thing I've found already, OpenBSD (and I believe NetBSD/Solaris) will require slightly different structure for sockaddr (specifically they expect RawSockAddr layout, instead of SockAddrDatalink). Also it appears at least with OpenBSD some of the other raw_bsd.go functionality is not working, so far I've confirmed BIOCSSEESENT is non-existant in the OpenBSD C headers nor is it in the Golang syscall/unix packages. FreeBSD aliases BIOCSDIRECTION to BIOCSSEESENT, however BIOCSDIRECTION is also not in OpenBSD. The OpenBSD equivalent appears to be BIOCSDIRFILT.

I've been spinning up FreeBSD and OpenBSD dev labs to validate all these changes, if I can muster i'll do the same for Solaris and NetBSD. At this time it seems we will have to stub out some of these functions into OS specific files (setsockopts_openbsd.go etc?)